Src/AutoNSubstitute/NSubstituteVirtualMethodsCommand.cs

```diff
@@ -172,40 +174,50 @@ public ISpecimenContext Context
                 get { return context; }
             }
 
-            public static void ReturnsForAnyArgs<T>(T value, Func<CallInfo, T> returnThis)
-            {
-                ReturnsForAnyArgsMethodInfo.Invoke(new object[] { value, returnThis });
-            }
-
             public static void Returns<T>(T value, Func<CallInfo, T> returnThis)
             {
                 ReturnsMethodInfo.Invoke(new object[] { value, returnThis });
             }
 
-            public void ReturnsUsingContext<T>(MethodInfo methodInfo)
+            public void ReturnsUsingContext(MethodInfo methodInfo)
             {
-                InvokeMethod(methodInfo);
-                ReturnsForAnyArgs<T>(default(T), callInfo =>
-                {
-                    var value = new Lazy<T>(() => (T)Context.Resolve(typeof(T)));
-                    object[] arguments = callInfo.Args();
-                    ReturnsFixedValue(methodInfo, callInfo, value);
-                    var returnValue = value.Value;
-                    InvokeMethod(methodInfo, arguments);
+                Substitute
+                    .WhenForAnyArgs(_ => InvokeMethod(methodInfo))
+                    .Do(callInfo =>
+                    {
+                        var arguments = callInfo.Args();
+                        var call = Tuple.Create(methodInfo, arguments.ToArray());
+                        if (methodCalls.Any(
+                            x =>
+                                call.Item1 == x.Item1 &&
+                                call.Item2.SequenceEqual(x.Item2)))
+                            return;
+                        methodCalls.Add(call);
+
+                        var value = Resolve(methodInfo.ReturnType);
+                        if (value is OmitSpecimen)
+                            return;
+
+                        ReturnsFixedValue(methodInfo, value);
+                        InvokeMethod(methodInfo, arguments);
+                    });
+            }
 
-                    return returnValue;
-                });
+            private object Resolve(Type type)
+            {
+                return Task.Factory
+                    .StartNew(() => Context.Resolve(type))
+                    .Result;
```

Comment:
This is needed when you have a method returning a circular reference like ``IInterfaceWithCircularDependency`` or ``IInterfaceWithCircularMethod``.

NSubstitute uses a static property ``SubstitutionContext.Current`` to get and set the state of its substitutes. However, it uses ThreadLocal fields so one thread does not interfere in another. For this reason, I have to force the ``Context.Resolve`` to run in another thread, so in cases where you have a circular reference, NSubstitute sets up the methods correctly.

This change has become needed because now I have to resolve the value, to check if it's an instance of ``OmitSpecimen``, before I set up the method. Before, the value was lazy and it was only evaluated after I had set up the method.